developers/plugin-architecture/write-custom-plugins/README.md

```diff
@@ -1,7 +1,53 @@
 # Write Custom Plugins
 
-Pinot allows user to easily create and use new plugins to support custom needs. The developer needs to implement a few standard interfaces specific to the the type of plugins in Java.
+Pinot allows user to easily create and use new plugins to support custom needs. The developer needs to implement a few standard interfaces specific to the type of plugins in Java.
 
-Once the code is ready, you can put the complete JAR file in pinot `/plugins` directory. All the classes in plugins directory are loaded at pinot's startup.\
-\
-We also encourage users to shaded commonly used dependencies such as `guava`, `jackson`&#x20;
+Once the code is ready, you can put the complete JAR file in pinot `/plugins` directory. All the classes in plugins directory are loaded at pinot's startup.
+
+As of Apache Pinot 1.3.0 there's an additional support for a new structure:
+
+     /plugins
+        /my-custom-plugin
+           /classes/...
+           /com-2.3.1.jar
+           /foo-1.2.1.jar
+           /bar-3.1.1.jar
+           /pinot-plugin.properties
+
+This way there's no need to shade classes anymore, because Pinot will ensure isolation and proper ordered loading of these classes. 
+Pinot will only expose the classes in pinot-spi.
+The `classes`-directory will contain custom compiled Java-code of the plugin.
```

Comment:
Do we actually need this `classes` directory? What is the difference between including the plugin classes in the `classes` directory compared to including an extra jar like `/my-custom-plugin.jar`?

Reply:
the code makes sure the classes is the first on the "classpath". If it is just a jar, you cannot do that. Ideally you want to have the exact order in which this code was compiled to prevent awkward edge cases.

Reply:
But the order is:

1. Classes from baseClassLoader
2. Classes in `my-custom-plugin/classes
3. Jars in `my-custom-plugin/*.jar`
4. Classes in package (and subpackages) `org.apache.pinot.spi`.
5. Classes from `importFrom`
6. Classes from `parent`

Given classes in `my-custom-plugin/classes` and `my-custom-plugin/*.jar` are controlled by the plugin developer, it should be safe (or at least their responsibility) to place plugin classes as a jar. Am I right?

Reply:
That's almost correct, and it is all in the details. There's one trick you can do to prevent shading: add the source of that class to your own codebase and adjust it. Maven will compile it and because Maven controls the order of classpath entries, this works. That said it is important to have the plugin classes as the very first on the classloader.
The assembly descriptor takes care of that.
There is another option: use the plugin.jar instead of classes, and specify the classpath in the pinot-plugin.properties. This way the order will be exactly the same as Maven uses it.

Reply:
From https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/File.html#list() 

> There is no guarantee that the name strings in the resulting array will appear in any specific order; they are not, in particular, guaranteed to appear in alphabetical order.